File: crates/dry_run_core/src/query/advise.rs
use serde::{Deserialize, Serialize};

use super::plan::PlanNode;
use super::suggest::{self, IndexSuggestion};
use crate::error::Result;
use crate::jit;
use crate::schema::{self, AnnotatedSchema, ColumnStats, QualifiedName};
use crate::version::PgVersion;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Advice {
    pub issue: String,
    pub severity: String,
    pub table: Option<String>,
    pub recommendation: String,
    pub ddl: Option<String>,
    pub version_note: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdviseResult {
    pub advice: Vec<Advice>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub index_suggestions: Vec<IndexSuggestion>,
}

// Top-level advise pass — walks the plan tree and emits per-node advice.
//
// Takes the annotated view rather than a raw `&SchemaSnapshot` because
// the per-node refinements (selectivity hints, partial-index suggestions,
// per-replica seq_scan breakdown) all need planner column stats and
// activity counters. Without those, advise still works — it just
// degrades to "DDL-only" recommendations.
pub fn advise(
    plan: &PlanNode,
    annotated: &AnnotatedSchema<'_>,
    pg_version: Option<&PgVersion>,
) -> Vec<Advice> {
    let mut advice = Vec::new();
    walk_for_advice(plan, annotated, pg_version, &mut advice);
    advice
}

// Full advise pass: plan-based advice + optional index suggestions via static SQL analysis.
// Works without a live DB when `plan` is None — falls back to query-structure analysis only.
pub fn advise_with_index_suggestions(
    sql: &str,
    plan: Option<&PlanNode>,
    annotated: &AnnotatedSchema<'_>,
    pg_version: Option<&PgVersion>,
    include_index_suggestions: bool,
) -> Result<AdviseResult> {
    let advice = match plan {
        Some(p) => advise(p, annotated, pg_version),
        None => Vec::new(),
    };

    let index_suggestions = if include_index_suggestions {
        // suggest_index reads `reltuples` for size cutoffs — pass the
        // annotated view so it has access to planner sizing.
        suggest::suggest_index(sql, annotated, plan, pg_version)?
    } else {
        Vec::new()
    };

    Ok(AdviseResult {
        advice,
        index_suggestions,
    })
}

fn walk_for_advice(
    node: &PlanNode,
    annotated: &AnnotatedSchema<'_>,
    pg_version: Option<&PgVersion>,
    advice: &mut Vec<Advice>,
) {
    advise_seq_scan(node, annotated, pg_version, advice);
    advise_nested_loop_seq_scan(node, pg_version, advice);
    advise_sort(node, pg_version, advice);
    advise_cte(node, advice);

    for child in &node.children {
        walk_for_advice(child, annotated, pg_version, advice);
    }
}

fn advise_seq_scan(
    node: &PlanNode,
    annotated: &AnnotatedSchema<'_>,
    pg_version: Option<&PgVersion>,
    advice: &mut Vec<Advice>,
) {
    if node.node_type != "Seq Scan" {
        return;
    }
    let table_name = match &node.relation_name {
        Some(n) => n,
        None => return,
    };
    if node.plan_rows < 10_000.0 {
        return;
    }

    let schema_name = node.schema.as_deref().unwrap_or("public");
    let qualified = format!("{schema_name}.{table_name}");
    let qn = QualifiedName::new(schema_name, table_name);

    let table = annotated
        .schema
        .tables
        .iter()
        .find(|t| t.name == *table_name && t.schema == schema_name);

    let filter_col = node
        .filter
        .as_ref()
        .and_then(|f| extract_column_from_filter(f));

    let has_index = if let (Some(table), Some(col)) = (&table, &filter_col) {
        table
            .indexes
            .iter()
            .any(|idx| idx.columns.first().map(|c| c.as_str()) == Some(col.as_str()))
    } else {
        false
    };

    if has_index {
        advice.push(Advice {
            issue: format!(
                "sequential scan on '{qualified}' (~{} rows) despite existing index",
                node.plan_rows as i64
            ),
            severity: "info".into(),
            table: Some(qualified),
            recommendation:
                "Run ANALYZE to update statistics. The planner may correctly prefer a seq scan if selectivity is low."
                    .into(),
            ddl: Some(format!("ANALYZE {schema_name}.{table_name};")),
            version_note: None,
        });
        return;
    }

    let (ddl, recommendation) = if let Some(filter_col_name) = &filter_col {
        let col_obj = table.and_then(|t| t.columns.iter().find(|c| c.name == *filter_col_name));
        let col_type = col_obj.map(|c| c.type_name.as_str()).unwrap_or("unknown");
        // Column stats live in the planner snapshot, keyed by qualified
        // table name + column name. Returns None if there's no planner
        // capture yet — in which case we fall back to non-stats advice.
        let col_stats = annotated.column_stats(&qn, filter_col_name);

        let (idx_type, rec) = suggest_index_type(&qualified, col_type, filter_col_name);
        let mut recommendation = rec;

        // Stats-aware refinements — only meaningful when we actually have
        // column stats. The plan's row estimate is the floor; if planner
        // sizing reports more rows than the plan rows estimate (which can
        // happen on stale plan estimates), prefer the larger number.
        if col_stats.is_some() {
            let mut table_rows = node.plan_rows;
            if let Some(rt) = annotated.reltuples(&qn)
                && rt > table_rows
            {
                table_rows = rt;
            }
            recommendation.push_str(&stats_aware_advice(col_stats, filter_col_name, table_rows));
        }

        let idx_name = format!("idx_{table_name}_{filter_col_name}");

        // Prefer a partial index for high-null or skewed columns — a tiny
        // selective index is much cheaper than a full one when most rows
        // would never match the predicate. Falls through to a plain
        // CREATE INDEX when stats aren't available.
        let null_frac = col_stats.and_then(|s| s.null_frac).unwrap_or(0.0);
        let ddl = if null_frac > 0.5 {
            format!(
                "CREATE INDEX CONCURRENTLY {idx_name} ON {schema_name}.{table_name} USING {idx_type}({filter_col_name}) WHERE {filter_col_name} IS NOT NULL;"
            )
        } else if let Some(stats) = col_stats
            && let Some((dominant, _freq)) = schema::has_skewed_distribution(stats, 0.5)
        {
            format!(
                "CREATE INDEX CONCURRENTLY {idx_name} ON {schema_name}.{table_name} USING {idx_type}({filter_col_name}) WHERE {filter_col_name} != '{dominant}';"
            )
        } else {
            format!(
                "CREATE INDEX CONCURRENTLY {idx_name} ON {schema_name}.{table_name} USING {idx_type}({filter_col_name});"
            )
        };

        (Some(ddl), recommendation)
    } else {
        (
            None,
            "Add an index on the filtered column(s) to avoid sequential scan.".into(),
        )
    };

    let mut full_recommendation = recommendation;

    // Per-node breakdown — surfaces "this replica is doing the unindexed
    // work, the others aren't" patterns. Empty when we only have one node
    // (or none); skipping the note in that case avoids noise.
    let per_node = annotated.seq_scan_per_node(&qn);
    if per_node.len() >= 2 {
        let total: i64 = per_node.iter().map(|(_, v)| *v).sum();
        let parts: Vec<String> = per_node
            .iter()
            .map(|(src, v)| format!("{src}: {v}"))
            .collect();
        full_recommendation.push_str(&format!(
            "\n\nNote: across {} nodes, seq_scan totals {} ({}). \
             Check if specific replicas are serving unindexed query patterns.",
            per_node.len(),
            total,
            parts.join(", ")
        ));
    }

    advice.push(Advice {
        issue: format!(
            "sequential scan on '{qualified}' (~{} rows)",
            node.plan_rows as i64
        ),
        severity: "warning".into(),
        table: Some(qualified),
        recommendation: full_recommendation,
        ddl,
        version_note: version_note_for_index(pg_version),
    });
}

fn advise_nested_loop_seq_scan(
    node: &PlanNode,
    pg_version: Option<&PgVersion>,
    advice: &mut Vec<Advice>,
) {
    if node.node_type != "Nested Loop" {
        return;
    }

    let inner = match node.children.get(1) {
        Some(child) if child.node_type == "Seq Scan" && child.plan_rows > 100.0 => child,
        _ => return,
    };

    let table_name = inner.relation_name.as_deref().unwrap_or("unknown");
    let schema_name = inner.schema.as_deref().unwrap_or("public");
    let qualified = format!("{schema_name}.{table_name}");

    let filter_col = inner
        .filter
        .as_ref()
        .and_then(|f| extract_column_from_filter(f));

    let ddl = filter_col.as_ref().map(|col| {
        format!(
            "CREATE INDEX CONCURRENTLY idx_{table_name}_{col} ON {schema_name}.{table_name}({col});"
        )
    });

    advice.push(Advice {
        issue: format!(
            "nested loop with sequential scan on inner side '{qualified}' (~{} rows per loop)",
            inner.plan_rows as i64
        ),
        severity: "warning".into(),
        table: Some(qualified),
        recommendation:
            "Add an index on the join/filter column of the inner table to convert the seq scan to an index scan."
                .into(),
        ddl,
        version_note: version_note_for_index(pg_version),
    });
}

fn advise_sort(node: &PlanNode, pg_version: Option<&PgVersion>, advice: &mut Vec<Advice>) {
    if node.node_type != "Sort" || node.plan_rows < 10_000.0 {
        return;
    }

    let sort_keys = match &node.sort_key {
        Some(keys) if !keys.is_empty() => keys,
        _ => return,
    };

    let table_info = find_table_in_subtree(node);
    let (schema_name, table_name) = match &table_info {
        Some((s, t)) => (s.as_str(), t.as_str()),
        None => return,
    };
    let qualified = format!("{schema_name}.{table_name}");

    let first_key = sort_keys[0]
        .split_whitespace()
        .next()
        .unwrap_or(&sort_keys[0]);

    let ddl = format!(
        "CREATE INDEX CONCURRENTLY idx_{table_name}_{first_key} ON {schema_name}.{table_name}({});",
        sort_keys.join(", ")
    );

    advice.push(Advice {
        issue: format!(
            "sort on ~{} rows (keys: {})",
            node.plan_rows as i64,
            sort_keys.join(", ")
        ),
        severity: "info".into(),
        table: Some(qualified),
        recommendation: "Consider an index matching the sort order to avoid an explicit sort step."
            .into(),
        ddl: Some(ddl),
        version_note: version_note_for_index(pg_version),
    });
}

// Build a recommendation suffix grounded in column stats — selectivity,
// dominant-value skew, null fraction, physical correlation. Returns an
// empty string when no stats are available, which lets the caller stitch
// it on unconditionally without a `match`.
fn stats_aware_advice(stats: Option<&ColumnStats>, filter_col: &str, table_rows: f64) -> String {
    let stats = match stats {
        Some(s) => s,
        None => return String::new(),
    };
    let mut parts = Vec::new();

    // Selectivity — the fraction of rows a value-equality predicate is
    // expected to match. Low cardinality (≤ 5 distinct values) → high
    // selectivity → poor index usefulness; we call that out explicitly.
    let sel = schema::column_selectivity(Some(stats), table_rows);
    if let Some(nd) = stats.n_distinct {
        if nd > 0.0 && nd <= 5.0 {
            parts.push(format!(
                "\nColumn '{}' has only {:.0} distinct values, so a full index has poor selectivity ({:.0}% of rows per value).",
                filter_col, nd, sel * 100.0
            ));
        } else if nd > 0.0 && nd <= 20.0 {
            parts.push(format!(
                "\nColumn '{}' has {} distinct values (selectivity ~{:.1}%).",
                filter_col,
                nd as i64,
                sel * 100.0
            ));
        }
    }

    // skew detection
    if let Some((dominant, freq)) = schema::has_skewed_distribution(stats, 0.5) {
        parts.push(format!(
            "Value '{}' dominates at ~{:.0}%. A partial index excluding it would be much smaller and faster.",
            dominant, freq * 100.0
        ));
    }

    // high null fraction
    if let Some(nf) = stats.null_frac
        && nf > 0.5
    {
        let null_rows = (nf * table_rows) as i64;
        parts.push(format!(
                "Column is {:.0}% NULL (~{} rows). Use a partial index WHERE {} IS NOT NULL to index only the non-null rows.",
                nf * 100.0, null_rows, filter_col
            ));
    }

    // correlation warning for range scans
    if let Some(c) = stats.correlation
        && c > -0.3
        && c < 0.3
        && table_rows > 10_000.0
    {
        parts.push(format!(
                "Physical ordering is random (correlation: {:.2}); index range scans will cause random I/O.",
                c
            ));
    }

    parts.join(" ")
}

fn advise_cte(node: &PlanNode, advice: &mut Vec<Advice>) {
    if node.node_type != "CTE Scan" {
        return;
    }
    let cte_name = match &node.cte_name {
        Some(n) => n,
        None => return,
    };
    let rows = node.plan_rows as i64;
    if rows < 1000 {
        return;
    }
    let e = jit::cte_materialized(cte_name, rows);
    advice.push(Advice {
        issue: format!("materialized CTE '{cte_name}' (~{rows} rows)"),
        severity: "info".into(),
        table: None,
        recommendation: format!("{}\n{}", e.reason, e.fix),
        ddl: None,
        version_note: None,
    });
}

// helpers

fn extract_column_from_filter(filter: &str) -> Option<String> {
    let trimmed = filter.trim().trim_start_matches('(').trim_end_matches(')');
    let first_token = trimmed.split_whitespace().next()?;
    let col = first_token.rsplit('.').next().unwrap_or(first_token);
    if col.chars().all(|c| c.is_alphanumeric() || c == '_') && !col.is_empty() {
        Some(col.to_string())
    } else {
        None
    }
}

fn suggest_index_type(table: &str, col_type: &str, col_name: &str) -> (&'static str, String) {
    let ct = col_type.to_lowercase();
    if ct == "jsonb" || ct == "tsvector" {
        let e = jit::suggest_gin(table, col_name, col_type);
        let rec = match &e.note {
            Some(note) => format!("{}\n{note}", e.reason),
            None => e.reason,
        };
        return ("gin", rec);
    }
    if ct.contains("geometry")
        || ct.contains("geography")
        || ct.contains("range")
        || ct == "tsrange"
        || ct == "daterange"
        || ct == "int4range"
    {
        let e = jit::suggest_gist(table, col_name, col_type);
        return ("gist", e.reason);
    }
    (
        "btree",
        format!("Add a B-tree index on '{col_name}' for equality/range lookups."),
    )
}

fn version_note_for_index(pg_version: Option<&PgVersion>) -> Option<String> {
    let ver = pg_version?;
    if ver.major >= 13 {
        Some("PG 13+: B-tree deduplication is enabled by default, reducing index size for low-cardinality columns.".into())
    } else if ver.major >= 11 {
        Some("PG 11+: Use INCLUDE for covering indexes to enable index-only scans.".into())
    } else {
        None
    }
}

fn find_table_in_subtree(node: &PlanNode) -> Option<(String, String)> {
    if let (Some(schema), Some(table)) = (&node.schema, &node.relation_name) {
        return Some((schema.clone(), table.clone()));
    }
    for child in &node.children {
        if let Some(result) = find_table_in_subtree(child) {
            return Some(result);
        }
    }
    None
}

#[cfg(test)]
#[path = "advise_tests.rs"]
mod tests;

File: internal/query/advise.go
package query

import (
	"fmt"
	"strings"

	"github.com/boringsql/dryrun/internal/dryrun"
	"github.com/boringsql/dryrun/internal/schema"
	"github.com/boringsql/dryrun/pkg/bloat"
	"github.com/boringsql/dryrun/pkg/jit"
)

type Advice struct {
	Issue            string            `json:"issue"`
	Severity         string            `json:"severity"`
	Table            *string           `json:"table,omitempty"`
	Recommendation   string            `json:"recommendation"`
	DDL              *string           `json:"ddl,omitempty"`
	VersionNote      *string           `json:"version_note,omitempty"`
	IndexSuggestions []IndexSuggestion `json:"index_suggestions,omitempty"`
}

// Walks plan tree, with per-node seq_scan breakdown when merged activity present
func Advise(plan *PlanNode, a *schema.AnnotatedSchema, pgVersion *dryrun.PgVersion) []Advice {
	var advice []Advice
	snap := a.Schema
	walkForAdvice(plan, a, pgVersion, &advice)

	if suggestions, err := SuggestIndex("", snap, plan, pgVersion); err == nil && len(suggestions) > 0 {
		for i := range advice {
			if advice[i].Table == nil {
				continue
			}
			for _, s := range suggestions {
				if s.Table == *advice[i].Table {
					advice[i].IndexSuggestions = append(advice[i].IndexSuggestions, s)
				}
			}
		}
	}

	if a.Merged != nil {
		for i := range advice {
			if advice[i].Table != nil && strings.Contains(advice[i].Issue, "sequential scan") {
				breakdown := perNodeBreakdown(a, *advice[i].Table)
				if breakdown != "" {
					advice[i].Recommendation += "\n\nPer-node breakdown:\n" + breakdown
				}
			}
		}
	}

	return advice
}

func perNodeBreakdown(a *schema.AnnotatedSchema, qualified string) string {
	if a == nil || a.Merged == nil {
		return ""
	}
	parts := strings.SplitN(qualified, ".", 2)
	if len(parts) != 2 {
		return ""
	}
	q := schema.QualifiedName{Schema: parts[0], Name: parts[1]}

	var lines []string
	for _, n := range a.Merged.Nodes {
		for _, ts := range n.Tables {
			if ts.Table == q {
				lines = append(lines, fmt.Sprintf("  %s: seq_scan=%d, idx_scan=%d", n.Node.Source, ts.Activity.SeqScan, ts.Activity.IdxScan))
			}
		}
	}
	return strings.Join(lines, "\n")
}

func walkForAdvice(node *PlanNode, a *schema.AnnotatedSchema, pgVersion *dryrun.PgVersion, advice *[]Advice) {
	adviseSeqScan(node, a, pgVersion, advice)
	adviseNestedLoopSeqScan(node, pgVersion, advice)
	adviseSort(node, a.Schema, pgVersion, advice)
	adviseIndexScanBloat(node, a, advice)
	adviseCTE(node, advice)

	for i := range node.Children {
		walkForAdvice(&node.Children[i], a, pgVersion, advice)
	}
}

func adviseSeqScan(node *PlanNode, a *schema.AnnotatedSchema, pgVersion *dryrun.PgVersion, advice *[]Advice) {
	if node.NodeType != "Seq Scan" || node.RelationName == nil || node.PlanRows < 10_000 {
		return
	}

	tableName := *node.RelationName
	schemaName := "public"
	if node.Schema != nil {
		schemaName = *node.Schema
	}
	qualified := schemaName + "." + tableName
	qual := schema.QualifiedName{Schema: schemaName, Name: tableName}

	var table *schema.Table
	for i := range a.Schema.Tables {
		if a.Schema.Tables[i].Name == tableName && a.Schema.Tables[i].Schema == schemaName {
			table = &a.Schema.Tables[i]
			break
		}
	}

	filterCol := extractColumnFromFilter(node.Filter)

	var matchingIdx *schema.Index
	if table != nil && filterCol != "" {
		for i := range table.Indexes {
			if len(table.Indexes[i].Columns) > 0 && table.Indexes[i].Columns[0] == filterCol {
				matchingIdx = &table.Indexes[i]
				break
			}
		}
	}

	if matchingIdx != nil {
		if sz := a.IndexSizingFor(qual, matchingIdx.Name); sz != nil {
			if est, ok := bloat.EstimateIndexBloat(*sz, matchingIdx.Columns, matchingIdx.IncludeColumns, *table, matchingIdx.IndexType); ok && est.BloatRatio > 3.0 {
				*advice = append(*advice, Advice{
					Issue:          fmt.Sprintf("sequential scan on '%s' (~%d rows) - index '%s' exists but appears bloated (%.1fx)", qualified, int64(node.PlanRows), matchingIdx.Name, est.BloatRatio),
					Severity:       "warning",
					Table:          strp(qualified),
					Recommendation: fmt.Sprintf("Index '%s' is estimated at %.1fx bloat. Rebuild it to restore accurate planner cost estimates.", matchingIdx.Name, est.BloatRatio),
					DDL:            strp(fmt.Sprintf("REINDEX CONCURRENTLY %s;", matchingIdx.Name)),
				})
				return
			}
		}

		ddl := fmt.Sprintf("ANALYZE %s.%s;", schemaName, tableName)
		*advice = append(*advice, Advice{
			Issue:          fmt.Sprintf("sequential scan on '%s' (~%d rows) despite existing index", qualified, int64(node.PlanRows)),
			Severity:       "info",
			Table:          strp(qualified),
			Recommendation: "Run ANALYZE to update statistics. The planner may correctly prefer a seq scan if selectivity is low.",
			DDL:            strp(ddl),
		})
		return
	}

	var ddl *string
	recommendation := "Add an index on the filtered column(s) to avoid sequential scan."
	if filterCol != "" {
		var col *schema.Column
		if table != nil {
			for i := range table.Columns {
				if table.Columns[i].Name == filterCol {
					col = &table.Columns[i]
					break
				}
			}
		}
		var colType string
		if col != nil {
			colType = col.TypeName
		}
		idxType, rec := suggestIndexType(qualified, colType, filterCol)
		recommendation = rec

		colStats := a.ColumnStats(qual, filterCol)
		if col != nil && colStats != nil {
			tableRows := node.PlanRows
			if sz := a.SizingFor(qual); sz != nil && sz.Reltuples > tableRows {
				tableRows = sz.Reltuples
			}
			recommendation += statsAwareAdvice(colStats, filterCol, tableRows)
		}

		idxName := fmt.Sprintf("idx_%s_%s", tableName, filterCol)

		if colStats != nil && colStats.NullFrac != nil && *colStats.NullFrac > 0.5 {
			ddl = strp(fmt.Sprintf("CREATE INDEX CONCURRENTLY %s ON %s.%s USING %s(%s) WHERE %s IS NOT NULL;",
				idxName, schemaName, tableName, idxType, filterCol, filterCol))
		} else if colStats != nil {
			if dominant, _, skewed := schema.HasSkewedDistribution(colStats, 0.5); skewed {
				ddl = strp(fmt.Sprintf("CREATE INDEX CONCURRENTLY %s ON %s.%s USING %s(%s) WHERE %s != '%s';",
					idxName, schemaName, tableName, idxType, filterCol, filterCol, dominant))
			} else {
				ddl = strp(fmt.Sprintf("CREATE INDEX CONCURRENTLY %s ON %s.%s USING %s(%s);",
					idxName, schemaName, tableName, idxType, filterCol))
			}
		} else {
			ddl = strp(fmt.Sprintf("CREATE INDEX CONCURRENTLY %s ON %s.%s USING %s(%s);",
				idxName, schemaName, tableName, idxType, filterCol))
		}
	}

	*advice = append(*advice, Advice{
		Issue:          fmt.Sprintf("sequential scan on '%s' (~%d rows)", qualified, int64(node.PlanRows)),
		Severity:       "warning",
		Table:          strp(qualified),
		Recommendation: recommendation,
		DDL:            ddl,
		VersionNote:    versionNoteForIndex(pgVersion),
	})
}

func adviseNestedLoopSeqScan(node *PlanNode, pgVersion *dryrun.PgVersion, advice *[]Advice) {
	if node.NodeType != "Nested Loop" || len(node.Children) < 2 {
		return
	}
	inner := &node.Children[1]
	if inner.NodeType != "Seq Scan" || inner.PlanRows <= 100 {
		return
	}

	tableName := "unknown"
	if inner.RelationName != nil {
		tableName = *inner.RelationName
	}
	schemaName := "public"
	if inner.Schema != nil {
		schemaName = *inner.Schema
	}
	qualified := schemaName + "." + tableName

	filterCol := extractColumnFromFilter(inner.Filter)
	var ddl *string
	if filterCol != "" {
		ddl = strp(fmt.Sprintf("CREATE INDEX CONCURRENTLY idx_%s_%s ON %s.%s(%s);",
			tableName, filterCol, schemaName, tableName, filterCol))
	}

	*advice = append(*advice, Advice{
		Issue:          fmt.Sprintf("nested loop with sequential scan on inner side '%s' (~%d rows per loop)", qualified, int64(inner.PlanRows)),
		Severity:       "warning",
		Table:          strp(qualified),
		Recommendation: "Add an index on the join/filter column of the inner table to convert the seq scan to an index scan.",
		DDL:            ddl,
		VersionNote:    versionNoteForIndex(pgVersion),
	})
}

func adviseSort(node *PlanNode, _ *schema.SchemaSnapshot, pgVersion *dryrun.PgVersion, advice *[]Advice) {
	if node.NodeType != "Sort" || node.PlanRows < 10_000 || len(node.SortKey) == 0 {
		return
	}

	tableInfo := findTableInSubtree(node)
	if tableInfo == nil {
		return
	}
	schemaName, tableName := tableInfo[0], tableInfo[1]
	qualified := schemaName + "." + tableName

	firstKey := strings.Fields(node.SortKey[0])[0]
	ddl := fmt.Sprintf("CREATE INDEX CONCURRENTLY idx_%s_%s ON %s.%s(%s);",
		tableName, firstKey, schemaName, tableName, strings.Join(node.SortKey, ", "))

	*advice = append(*advice, Advice{
		Issue:          fmt.Sprintf("sort on ~%d rows (keys: %s)", int64(node.PlanRows), strings.Join(node.SortKey, ", ")),
		Severity:       "info",
		Table:          strp(qualified),
		Recommendation: "Consider an index matching the sort order to avoid an explicit sort step.",
		DDL:            strp(ddl),
		VersionNote:    versionNoteForIndex(pgVersion),
	})
}

func extractColumnFromFilter(filter *string) string {
	if filter == nil {
		return ""
	}
	trimmed := strings.TrimSpace(*filter)
	trimmed = strings.TrimLeft(trimmed, "(")
	trimmed = strings.TrimRight(trimmed, ")")
	fields := strings.Fields(trimmed)
	if len(fields) == 0 {
		return ""
	}
	token := fields[0]
	if i := strings.LastIndex(token, "."); i >= 0 {
		token = token[i+1:]
	}
	for _, c := range token {
		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '_') {
			return ""
		}
	}
	if token == "" {
		return ""
	}
	return token
}

func suggestIndexType(table, colType, colName string) (string, string) {
	ct := strings.ToLower(colType)
	switch {
	case ct == "jsonb":
		e := jit.SuggestGIN(table, colName, colType)
		return "gin", e.Reason + "\n" + e.Note
	case ct == "tsvector":
		e := jit.SuggestGIN(table, colName, colType)
		return "gin", e.Reason
	case strings.Contains(ct, "geometry") || strings.Contains(ct, "geography"):
		e := jit.SuggestGiST(table, colName, colType)
		return "gist", e.Reason
	case strings.Contains(ct, "range") || ct == "tsrange" || ct == "daterange" || ct == "int4range":
		e := jit.SuggestGiST(table, colName, colType)
		return "gist", e.Reason
	default:
		return "btree", fmt.Sprintf("Add a B-tree index on '%s' for equality/range lookups.", colName)
	}
}

func versionNoteForIndex(pgVersion *dryrun.PgVersion) *string {
	if pgVersion == nil {
		return nil
	}
	if pgVersion.Major >= 13 {
		return strp("PG 13+: B-tree deduplication is enabled by default, reducing index size for low-cardinality columns.")
	}
	if pgVersion.Major >= 11 {
		return strp("PG 11+: Use INCLUDE for covering indexes to enable index-only scans.")
	}
	return nil
}

func adviseIndexScanBloat(node *PlanNode, a *schema.AnnotatedSchema, advice *[]Advice) {
	if node.IndexName == nil {
		return
	}
	if node.NodeType != "Index Scan" && node.NodeType != "Index Only Scan" {
		return
	}

	tableName := ""
	schemaName := "public"
	if node.RelationName != nil {
		tableName = *node.RelationName
	}
	if node.Schema != nil {
		schemaName = *node.Schema
	}
	if tableName == "" {
		return
	}
	qual := schema.QualifiedName{Schema: schemaName, Name: tableName}

	var table *schema.Table
	for i := range a.Schema.Tables {
		if a.Schema.Tables[i].Name == tableName && a.Schema.Tables[i].Schema == schemaName {
			table = &a.Schema.Tables[i]
			break
		}
	}
	if table == nil {
		return
	}

	indexName := *node.IndexName
	for _, idx := range table.Indexes {
		if idx.Name != indexName {
			continue
		}
		sz := a.IndexSizingFor(qual, indexName)
		if sz == nil {
			break
		}
		est, ok := bloat.EstimateIndexBloat(*sz, idx.Columns, idx.IncludeColumns, *table, idx.IndexType)
		if ok && est.BloatRatio > 3.0 {
			qualified := schemaName + "." + tableName
			*advice = append(*advice, Advice{
				Issue:          fmt.Sprintf("index '%s' on '%s' appears bloated (%.1fx) - cost estimates may be inflated", indexName, qualified, est.BloatRatio),
				Severity:       "info",
				Table:          strp(qualified),
				Recommendation: fmt.Sprintf("Rebuild index to improve cost accuracy: REINDEX CONCURRENTLY %s;", indexName),
				DDL:            strp(fmt.Sprintf("REINDEX CONCURRENTLY %s;", indexName)),
			})
		}
		break
	}
}

func statsAwareAdvice(s *schema.ColumnStats, filterCol string, tableRows float64) string {
	if s == nil {
		return ""
	}
	var parts []string

	sel := schema.ColumnSelectivity(s, tableRows)
	if s.NDistinct != nil {
		nd := *s.NDistinct
		if nd > 0 && nd <= 5 {
			parts = append(parts, fmt.Sprintf("\nColumn '%s' has only %.0f distinct values, so a full index has poor selectivity (%.0f%% of rows per value).", filterCol, nd, sel*100))
		} else if nd > 0 && nd <= 20 {
			parts = append(parts, fmt.Sprintf("\nColumn '%s' has %d distinct values (selectivity ~%.1f%%).", filterCol, int64(nd), sel*100))
		}
	}

	if dominant, freq, skewed := schema.HasSkewedDistribution(s, 0.5); skewed {
		parts = append(parts, fmt.Sprintf("Value '%s' dominates at ~%.0f%%. A partial index excluding it would be much smaller and faster.", dominant, freq*100))
	}

	if s.NullFrac != nil && *s.NullFrac > 0.5 {
		nullRows := int64(*s.NullFrac * tableRows)
		parts = append(parts, fmt.Sprintf("Column is %.0f%% NULL (~%d rows). Use a partial index WHERE %s IS NOT NULL to index only the non-null rows.", *s.NullFrac*100, nullRows, filterCol))
	}

	if s.Correlation != nil {
		c := *s.Correlation
		if c > -0.3 && c < 0.3 && tableRows > 10_000 {
			parts = append(parts, fmt.Sprintf("Physical ordering is random (correlation: %.2f); index range scans will cause random I/O.", c))
		}
	}

	return strings.Join(parts, " ")
}

func adviseCTE(node *PlanNode, advice *[]Advice) {
	if node.NodeType != "CTE Scan" || node.CTEName == nil {
		return
	}
	rows := int64(node.PlanRows)
	if rows < 1000 {
		return
	}
	cteName := *node.CTEName
	e := jit.CTEMaterialized(cteName, rows)
	*advice = append(*advice, Advice{
		Issue:          fmt.Sprintf("materialized CTE '%s' (~%d rows)", cteName, rows),
		Severity:       "info",
		Recommendation: e.Reason + "\n" + e.Fix,
	})
}

func findTableInSubtree(node *PlanNode) []string {
	if node.Schema != nil && node.RelationName != nil {
		return []string{*node.Schema, *node.RelationName}
	}
	for i := range node.Children {
		if result := findTableInSubtree(&node.Children[i]); result != nil {
			return result
		}
	}
	return nil
}
